### PR TITLE
[2201.9.x] Reorder the desugaring of type definitions before the desugaring of global variables and configurables in the desugar phase

### LINKED ISSUE #42829: [Bug]: When trying to create a Java String for a enum value a `TypeCastError` is thrown
### Description

> $subject

### Steps to Reproduce

Consider the following code segment:

```ballerina
import ballerina/jballerina.java as java;

public enum LogLevel {
    DEBUG,
    INFO,
    WARNING,
    ERROR,
    FATAL,
    OFF
}

public type CustomConfiguration record {
    LogLevel logLevel = OFF;
};

configurable CustomConfiguration customConfiguration = {};

public isolated client class SampleClient {
    private final readonly & LogLevel logLevel;

    public isolated function init() returns error? {
        self.logLevel = customConfiguration.logLevel;
        handle logHandle = java:fromString(self.logLevel);
    }
}

public function main() returns error? {
    SampleClient clientEp = check new;
}
```

Following runtime error is thrown when trying to run the above program:
```
error: {ballerina}TypeCastError {"message":"incompatible types: '()' cannot be cast to 'string'"}
	at ayesh.asb_issue_test.0.SampleClient:init(client.bal:25)
	   ayesh.asb_issue_test.0:main(client.bal:31)
```

### Affected Version(s)

Ballerina SL 2201.9.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)
- [#6573](https://github.com/ballerina-platform/ballerina-library/issues/6573)

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #42829 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -937,7 +937,7 @@ private BLangStatementExpression createIfElseFromConfigurable(BLangSimpleVariabl
         BLangStatementExpression stmtExpr = createStatementExpression(ifElse, resultVarRef);
         stmtExpr.setBType(configurableVar.getBType());
 
-        return rewrite(stmtExpr, initFunctionEnv);
+        return stmtExpr;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:940]: Can we add a comment here explaining why we are not rewriting here?

Usually when we desugar stuff we desugar completely.

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:940]: Desugaring here is not the correct approach. We don't desugar any expressions of global-level variables here. The expressions of global-level variables are desugared when visiting the init function.

Do we need to add a special comment here?

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:941]: https://github.com/ballerina-platform/ballerina-lang/pull/42853/commits/b1e9ce64742fe61c6530a488e6377ed611b64a6d

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -937,7 +937,8 @@ private BLangStatementExpression createIfElseFromConfigurable(BLangSimpleVariabl
         BLangStatementExpression stmtExpr = createStatementExpression(ifElse, resultVarRef);
         stmtExpr.setBType(configurableVar.getBType());
 
-        return rewrite(stmtExpr, initFunctionEnv);
+        // This statement expression is desugared when visiting the init function
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:940]:
```suggestion
        // This statement expression is desugared when it is visited in the init function
        // to allow desugared record default closures to be used as expected.
```

**Pull request merged**